### PR TITLE
 [MP Rewrite Part 1] Break out and clean up BallState 

### PR DESCRIPTION
**Pull request by kylestach**
This substantially cleans up the ball state code, and moves it from
`SystemState` (which we are moving away from in general) to `WorldState`
(which is restricted to physics-only information).

### CODE REVIEW

File: soccer/WorldState.hpp [hunk 1]
```diff
@@ -14,12 +20,80 @@ struct RobotState {
     Geometry2d::Twist velocity;
     RJ::Time timestamp;
     bool visible = false;
-    bool velocity_valid = false;
 };
 
+/**
+ * Our belief about the ball's current position and velocity.
```
Comment (oswinso) [inline soccer/WorldState.hpp]: `\brief`

File: soccer/WorldState.hpp [hunk 2]
```diff
@@ -14,12 +20,80 @@ struct RobotState {
...
+    RJ::Time timestamp;
+    bool visible = false;
+
+    /**
+     * Default constructor: an invalid ball with visible=false.
+     */
+    BallState() = default;
+
+    /**
+     * Construct a BallState with a valid estimate.
```
Comment (oswinso) [inline soccer/WorldState.hpp]: `\brief` for all of these?

File: soccer/WorldState.hpp [hunk 3]
```diff
@@ -14,12 +20,80 @@ struct RobotState {
...
+    /**
+     * Predict the ball's state at a particular instance in time.
+     * @param time the time at which to evaluate the ball's position.
+     * @return the ball's state.
+     */
+    [[nodiscard]] BallState predict_at(RJ::Time time) const;
+
+    /**
+     * Similar to \ref predict_at "predict_at(RJ::Time)", but for a duration in
+     * the future
```
Comment (oswinso) [inline soccer/WorldState.hpp]: Clarify by saying it's `timestamp + seconds`

File: soccer/WorldState.hpp [hunk 4]
```diff
@@ -1,7 +1,13 @@
...
 
-#include <Constants.hpp>
-#include <Geometry2d/Pose.hpp>
+#include "Constants.hpp"
+#include "Geometry2d/Pose.hpp"
+#include "planning/Instant.hpp"
+#include "planning/Trajectory.hpp"
+#include "time.hpp"
+
+// TODO(Kyle): Make this configurable
```
Comment (oswinso) [inline soccer/WorldState.hpp]: Issue number pls

File: soccer/gameplay/robocup-py.cpp [hunk 5]
```diff
@@ -904,21 +1007,20 @@ BOOST_PYTHON_MODULE(robocup) {
...
         .def_readwrite("is_penalty_kicker", &OurRobot::isPenaltyKicker)
-        .def_readwrite("is_ball_placer", &OurRobot::isBallPlacer)
-        .def("is_facing", &OurRobot::isFacing);
+        .def_readwrite("is_ball_placer", &OurRobot::isBallPlacer);
 
     class_<OpponentRobot, OpponentRobot*, std::shared_ptr<OpponentRobot>,
            bases<Robot>>("OpponentRobot", init<Context*, int>());
 
-    class_<Ball, std::shared_ptr<Ball>>("Ball", init<>())
+    class_<BallState, BallState*>("Ball", init<>())
```
Comment (oswinso) [inline soccer/gameplay/robocup-py.cpp:1023]: Why the change from `shared_ptr` to raw ptr?

**Review by oswinso — CHANGES_REQUESTED** [review]
A few small things

File: soccer/WorldState.hpp [hunk 6]
```diff
@@ -14,12 +20,80 @@ struct RobotState {
...
+     * Predict the stop time of the ball.
+     * @param out will be filled with the stopping position,
+     *  if it is not nullptr.
+     * @return the duration until the ball stops.
+     */
+    [[nodiscard]] RJ::Seconds query_stop_time(
+        Geometry2d::Point* out = nullptr) const;
+
+    /**
+     * Query the time before the ball goes a certain distance.
```
Comment (JNeiger) [inline soccer/WorldState.hpp]: Not: param and return descriptions. Specially the optional return

File: soccer/gameplay/robocup-py.cpp [hunk 7]
```diff
@@ -144,19 +170,18 @@ void OurRobot_move_to_end_vel(OurRobot* self, Geometry2d::Point* endPos,
...
-    if (to == nullptr) throw NullArgumentException("to");
+    if (to == nullptr) {
+        throw NullArgumentException("to");
+    }
     self->move(*to);
 }
 
-void OurRobot_settle(OurRobot* self) { self->settle(std::nullopt); }
-
-void OurRobot_settle_w_bounce(OurRobot* self, Geometry2d::Point* bounceTarget) {
```
Comment (JNeiger) [inline soccer/gameplay/robocup-py.cpp:153]: This removes the one touch functionality in the path planned layer. We'd have to revert back to the previous version, which I think just used move

File: soccer/planning/DynamicObstacle.hpp [hunk 8]
```diff
@@ -1,54 +1,13 @@
 #pragma once
-
-#include <Geometry2d/Point.hpp>
 #include <Geometry2d/Circle.hpp>
-#include "planning/MotionInstant.hpp"
-#include "Utils.hpp"
-#include "Geometry2d/Shape.hpp"
+#include <memory>
```
Comment (JNeiger) [inline soccer/planning/DynamicObstacle.hpp]: Is this include needed?

File: soccer/planning/DynamicObstacle.hpp [hunk 9]
```diff
@@ -1,54 +1,13 @@
...
-
-    // Radius = radius of obstacle
-    float getRadius() const { return radius; }
-
-    std::shared_ptr<Geometry2d::Circle> getStaticObstacle() const {
-        return staticObstacle;
-    }
+class Trajectory;
+struct DynamicObstacle {
+    DynamicObstacle(const Geometry2d::Circle& circle, const Trajectory* path)
```
Comment (JNeiger) [inline soccer/planning/DynamicObstacle.hpp]: Need some comments on what all this represents and how to use this

File: soccer/BallTest.cpp [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+TEST(BallState, QueryFar) {
+    RJ::Time start = RJ::now();
+    BallState state(Geometry2d::Point(0, 0), Geometry2d::Point(1, 0), start);
+
+    Geometry2d::Point actual;
+    RJ::Seconds t = state.query_seconds_near(Geometry2d::Point(5, 0), &actual);
+
+    // TODO: For some reason query_stop_time and query_seconds_to_dist give
```
Comment (JNeiger) [inline soccer/BallTest.cpp]: Issue #?

**Review by JNeiger — CHANGES_REQUESTED** [review]

File: soccer/gameplay/robocup-py.cpp (hunk 5 above)
Comment (kylestach) [reply soccer/gameplay/robocup-py.cpp:1023]: We no longer have a reason to use `BallState` with `shared_ptr` values anywhere (it's a tiny little value type, there's no reason to avoid copies).

File: soccer/WorldState.hpp (hunk 4 above)
Comment (kylestach) [reply soccer/WorldState.hpp]: #1498

File: soccer/gameplay/robocup-py.cpp (hunk 7 above)
Comment (kylestach) [reply soccer/gameplay/robocup-py.cpp:153]: Reverted.

File: soccer/planning/DynamicObstacle.hpp (hunk 9 above)
Comment (JNeiger) [reply soccer/planning/DynamicObstacle.hpp]: Circles have a center and radius, are there any specifics on the center?

**Review by JNeiger — COMMENTED** [review]
Minus the circle comment, this is good to merge

File: soccer/planning/DynamicObstacle.hpp (hunk 9 above)
Comment (kylestach) [reply soccer/planning/DynamicObstacle.hpp]: Clarified by making this only take a radius instead of a real circle object.

**Review by oswinso — APPROVED** [review]
lgtm

**Review by JNeiger — APPROVED** [review]
Lgtm

File: soccer/gameplay/plays/defense/DefendClear.py [hunk 11]
```diff
@@ -0,0 +1,35 @@
+import standard_play
```
Comment (ghost) [inline soccer/gameplay/plays/defense/DefendClear.py]: These plays files might have been added on accident. Is this supposed to be here?

File: soccer/WorldState.cpp [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+    // Otherwise, use t = (vf - vi) / a
+    return RJ::Seconds(speed - std::sqrt(vf_sq)) / kBallDecayConstant;
+}
+
+Planning::Trajectory BallState::make_trajectory() const {
+    using namespace Geometry2d;
+
+    // The trajectory interface fits cubic splines. Luckily, a cubic spline
+    // between two instants that can be connected by a constant acceleration
+    // (like we have here) will be, and so we can use this for our trajectory.
```
Comment (ghost) [inline soccer/WorldState.cpp:117]: This is pretty clever. nice.

**Review by ghost — CHANGES_REQUESTED** [review]

**Review by ghost — APPROVED** [review]
